executor/infoschema_reader.go

```diff
@@ -1426,6 +1427,125 @@ func (e *memtableRetriever) setDataFromTableConstraints(ctx sessionctx.Context,
 	e.rows = rows
 }
 
+// tableStorageStatsRetriever is used to read slow log data.
+type tableStorageStatsRetriever struct {
+	dummyCloser
+	table         *model.TableInfo
+	outputCols    []*model.ColumnInfo
+	retrieved     bool
+	initialized   bool
+	extractor     *plannercore.TableStorageStatsExtractor
+	initialTables []*initialTable
+	curTable      int
+	pdAddress     string
+	helper        *helper.Helper
+	stats         helper.PDRegionStats
+}
+
+func (e *tableStorageStatsRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		err := e.initialize(sctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(e.initialTables) == 0 || e.curTable >= len(e.initialTables) {
+		e.retrieved = true
+		return nil, nil
+	}
+
+	rows, err := e.setDataForTableStorageStats(sctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(e.outputCols) == len(e.table.Columns) {
+		return rows, nil
+	}
+	retRows := make([][]types.Datum, len(rows))
+	for i, fullRow := range rows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		retRows[i] = row
+	}
+	return retRows, nil
+}
+
+type initialTable struct {
+	db string
+	*model.TableInfo
+}
+
+func (e *tableStorageStatsRetriever) initialize(sctx sessionctx.Context) error {
+	is := infoschema.GetInfoSchema(sctx)
+	dbs := is.AllSchemas()
+	sort.Sort(infoschema.SchemasSorter(dbs))
+	schemas := e.extractor.TableSchema
+	tables := e.extractor.TableName
+
+	// Extract the tables to the initialTable.
+	for _, schema := range dbs {
```

Comment:
Could we replace it with the following code? Then we needn't use the `for` statement.
```
if len(schemas) != 0 {
    dbs, ok = is.SchemaByName(schema.Name)
}
```

Reply:
+1

Reply:
The for statement is necessary because the schemas may be multi. So i have made these change：
1. The user must specify the schema (If not specify the schema, it would traverse all the schema and tables to find the tables, which would cost much time. Also in the http api, the user must specify both schema and table), if only specify the schema and not specify table, it would return all tables information of the specified schema. 
2. Filter the sys or memory schema. select these schema would return null set.